### PR TITLE
Supercharge `TelegramObject.to_dict`

### PR DESCRIPTION
**Pull request by harshil21**
Moves the logic of `TO.to_dict` in subclasses to `TelegramObject` itself, which was mainly dealing with sequences and converting datetimes to timestamps.

### Checklist for PRs

- [ ] ~Added `.. versionadded:: version`, `.. versionchanged:: version` or `.. deprecated:: version` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)~
- [x] Created new or adapted existing unit tests
- [x] Documented code changes according to the [CSI standard](https://standards.mousepawmedia.com/en/stable/csi.html)
- [ ] ~Added new classes & modules to the docs and all suitable `__all__` s~

### CODE REVIEW

File: telegram/_bot.py [hunk 1]
```diff
@@ -3665,7 +3665,7 @@ async def edit_message_text(
         if inline_message_id:
             data["inline_message_id"] = inline_message_id
         if entities:
-            data["entities"] = [me.to_dict() for me in entities]
+            data["entities"] = [me.to_dict(recursive=True) for me in entities]
```
Comment (Bibo-Joshi) [inline telegram/_bot.py:3668]: Are the changes in `_bot` necessary?

File: telegram/_telegramobject.py [hunk 2]
```diff
@@ -327,6 +341,34 @@ def to_dict(self, recursive: bool = True) -> JSONDict:
             :obj:`dict`
         """
         out = self._get_attrs(recursive=recursive)
+
+        # Now we should convert TGObjects to dicts inside objects such as sequences, and convert
+        # datetimes to timestamps. This mostly eliminates the need for subclasses to override
+        # `to_dict`
+        for key, value in out.items():
+            if isinstance(value, Sequence) and not isinstance(value, str):
```
Comment (Bibo-Joshi) [inline telegram/_telegramobject.py:349]:
```suggestion
            if isinstance(value, Sequence) and not isinstance(value, (str, bytes)):
```

Strictly speaking the ` not isintance`  is not even necessary, as the `hasattr(item, "to_dict"` check will fail, but it's good to keep it so that we don't unnecessarily loop over things that we know are not TOs :D

File: telegram/_telegramobject.py [hunk 3]
```diff
@@ -327,6 +341,34 @@ def to_dict(self, recursive: bool = True) -> JSONDict:
...
+                val = []  # empty list to append our converted values to
+                for item in value:
+                    if hasattr(item, "to_dict"):
+                        val.append(item.to_dict(recursive=recursive))  # type: ignore[union-attr]
+                    # This branch is useful for e.g. List[List[PhotoSize|KeyboardButton]]
+                    elif isinstance(item, (tuple, list)):
```
Comment (Bibo-Joshi) [inline telegram/_telegramobject.py:355]: why not `Sequence` here?

On second thought, we do know that all the attributes of TO are only lists (or tuples after #3249), so checking for that (both here and above) might actually suffice 🤔 I might have been caught in the "ABC" spirit 😄

File: telegram/_telegramobject.py [hunk 4]
```diff
@@ -327,6 +341,34 @@ def to_dict(self, recursive: bool = True) -> JSONDict:
             :obj:`dict`
         """
         out = self._get_attrs(recursive=recursive)
+
+        # Now we should convert TGObjects to dicts inside objects such as sequences, and convert
+        # datetimes to timestamps. This mostly eliminates the need for subclasses to override
+        # `to_dict`
+        for key, value in out.items():
+            if isinstance(value, Sequence) and not isinstance(value, str):
+                val = []  # empty list to append our converted values to
```
Comment (Bibo-Joshi) [inline telegram/_telegramobject.py:350]:
```suggestion
                if len(value) == 0:
                    out[key] = value
                     continue
                     
                val = []  # empty list to append our converted values to
```

go, early returns, go! 🤡 That way you can also skip lines 365-366

File: tests/test_bot.py [hunk 5]
```diff
@@ -2998,7 +2998,10 @@ async def test_arbitrary_callback_data_pinned_message_reply_to_message(
         )
 
         message = Message(
-            1, None, None, reply_markup=bot.callback_data_cache.process_keyboard(reply_markup)
+            1,
+            dtm.datetime.utcnow(),
```
Comment (Bibo-Joshi) [inline tests/test_bot.py:3002]: can you pinpoint why passing an (admittably invalid) `None` here doesn't work anymore? 🤔

File: tests/test_telegramobject.py [hunk 6]
```diff
@@ -238,9 +238,9 @@ def test_pickle(self, bot):
         with pytest.raises(RuntimeError):
             unpickled.get_bot()  # There should be no bot when we pickle TGObjects
 
-        assert unpickled.chat == chat
+        assert unpickled.chat == chat, f"{unpickled.chat._id_attrs} != {chat._id_attrs}"
         assert unpickled.from_user == user
-        assert unpickled.date == date
+        assert unpickled.date == date, f"{unpickled.date} != {date}"
```
Comment (Bibo-Joshi) [inline tests/test_telegramobject.py:243]: these changes are not really related right? just double checking, we can keep them either way

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Nice, and it's not even fridy 🥳 Could you also extend the `to_dict` tests to make sure that the `recursive` setting now properly extends to the (nested) lists?

File: telegram/_bot.py (hunk 1 above)
Comment (harshil21) [reply telegram/_bot.py:3668]: no, I passed in the parameter cause you like it 😃

File: telegram/_telegramobject.py (hunk 3 above)
Comment (harshil21) [reply telegram/_telegramobject.py:355]: > why not `Sequence` here?

it is code from my initial attempt, before I figured I could use Sequence.

but okay, will revert to using tuple and list only

File: tests/test_bot.py (hunk 5 above)
Comment (harshil21) [reply tests/test_bot.py:3002]: it's because we did `data["date"] = to_timestamp(self.date)` in `Message.to_dict`. Where `to_timestamp` returned `None` if `self.date` was `None`. This means that `Message.de_json` would pass. But now we don't do that, so `Message.de_json` returns `KeyError: data["date"]`

File: tests/test_telegramobject.py (hunk 6 above)
Comment (harshil21) [reply tests/test_telegramobject.py:243]: yeah not related. They had failed during testing, so wanted to find out what the problem was.

**Comment by harshil21** [comment]
> Could you also extend the `to_dict` tests to make sure that the `recursive` setting now properly extends to the (nested) lists?

This is already tested in the classes which have those, e.g. [`test_to_dict` in `test_inlinekeyboardmarkup`](https://github.com/python-telegram-bot/python-telegram-bot/blob/rm-extra-to-dict/tests/test_inlinekeyboardmarkup.py#L137).

File: tests/test_bot.py (hunk 5 above)
Comment (Bibo-Joshi) [reply tests/test_bot.py:3002]: > But now we don't do that

Because `TO._get_attrs` skips `None` values, right? Okay, that sounds reasonable :)

Comment (harshil21) [reply tests/test_bot.py:3002]: > Because TO._get_attrs skips None values, right?

Precisely

**Review by Bibo-Joshi — APPROVED** [review]
DS failure is unrelated. When the tests pass, this LGTM :)

File: telegram/_passport/credentials.py [hunk 7]
```diff
@@ -450,15 +441,6 @@ class DataCredentials(_CredentialsBase):
     def __init__(self, data_hash: str, secret: str, *, api_kwargs: JSONDict = None):
         super().__init__(hash=data_hash, secret=secret, api_kwargs=api_kwargs)
 
-    def to_dict(self, recursive: bool = True) -> JSONDict:
-        """See :meth:`telegram.TelegramObject.to_dict`."""
-        data = super().to_dict(recursive=recursive)
-
-        del data["file_hash"]
-        del data["hash"]
```
Comment (harshil21) [inline telegram/_passport/credentials.py:458]: btw I hope this doesn't cause problems, the tests passed soo

Comment (Bibo-Joshi) [reply telegram/_passport/credentials.py:458]: mh, seems to be related to lines 429++, so should be good 🤔
